src/csharp/Grpc.Core/Interceptors/GenericInterceptor.cs

```diff
@@ -0,0 +1,449 @@
+#region Copyright notice and license
+
+// Copyright 2018 gRPC authors.
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+#endregion
+
+using System;
+using System.Threading;
+using System.Threading.Tasks;
+using Grpc.Core.Internal;
+
+namespace Grpc.Core.Interceptors
+{
+    /// <summary>
+    /// Provides a base class for generic interceptor implementations that raises
+    /// events and hooks to control the RPC lifecycle.
+    /// </summary>
+    public abstract class GenericInterceptor : Interceptor
+    {
+        /// <summary>
+        /// Provides hooks through which an invocation should be intercepted.
+        /// </summary>
+        public sealed class ClientCallArbitrator<TRequest, TResponse>
```

Comment:
I'm not super keen on the name "arbitrator", but I have to give it more thought (Are there any alternative names to consider?).

Reply:
How about `Controller`? I also considered `Observer` and `Listener`, but there are properties that let you override and control as it goes by, so it is not a passive object, necessarily.